### ISSUE TITLE
Repeating filename annotations when using CMake on Windows

### ISSUE THREAD

**Issue by peterjc123** [pytorch/pytorch#20686]
We are seeing filenames in `aten/src/ATen/native/cpu/` like the following cases:
```
Activation.cpp
Activation.cpp..cpp
Activation.cpp..cpp..cpp
Activation.cpp.AVX.cpp
Activation.cpp.AVX.cpp..cpp
Activation.cpp.AVX.cpp..cpp..cpp
Activation.cpp.AVX.cpp.AVX.cpp
UnaryOpsKernel.cpp.DEFAULT.cpp.DEFAULT.cpp.AVX2.cpp.AVX2.cpp
UnaryOpsKernel.cpp.DEFAULT.cpp.DEFAULT.cpp.DEFAULT.cpp.AVX2.cpp
UnaryOpsKernel.cpp.DEFAULT.cpp.AVX.cpp.DEFAULT.cpp.AVX2.cpp
```
They will slow down the build, and sometimes even make it hang if we don't clean the build cache and do the in-source build.

**Comment by peterjc123** [comment]
@ezyang May I know how are these source files generated so that I can fix this issue?

**Comment by peterjc123** [comment]
I confirm this issue is preventing the caffe2 build job on Windows using the new Windows AMI.

**Comment by ezyang** [comment]
This is related to generating extra copies of files so we can build them multiple times for AVX2.  I think this code is responsible for this:

```
  FOREACH(i RANGE ${NUM_CPU_CAPABILITY_NAMES})
    FOREACH(IMPL ${cpu_kernel_cpp_in})
      string(REPLACE "${CMAKE_CURRENT_LIST_DIR}/../aten/src/ATen/" "" NAME ${IMPL})
      LIST(GET CPU_CAPABILITY_NAMES ${i} CPU_CAPABILITY)
      SET(NEW_IMPL ${CMAKE_BINARY_DIR}/aten/src/ATen/${NAME}.${CPU_CAPABILITY}.cpp)
      CONFIGURE_FILE(${IMPL} ${NEW_IMPL} COPYONLY)
      SET(cpu_kernel_cpp ${NEW_IMPL} ${cpu_kernel_cpp}) # Create list of copies
      LIST(GET CPU_CAPABILITY_FLAGS ${i} FLAGS)
      IF(MSVC)
        SET(MACRO_FLAG "/DCPU_CAPABILITY=${CPU_CAPABILITY} /DCPU_CAPABILITY_${CPU_CAPABILITY}")
      ELSE(MSVC)
        SET(MACRO_FLAG "-DCPU_CAPABILITY=${CPU_CAPABILITY} -DCPU_CAPABILITY_${CPU_CAPABILITY}")
      ENDIF(MSVC)
      SET_SOURCE_FILES_PROPERTIES(${NEW_IMPL} PROPERTIES COMPILE_FLAGS "${FLAGS} ${MACRO_FLAG}")
    ENDFOREACH()
  ENDFOREACH()
  list(APPEND ATen_CPU_SRCS ${cpu_kernel_cpp})
```

in cmake/Codegen.cmake

It sounds like the problem is that this codegen should be pointed into build, not the source dir. @peterjc123, do you think this is enough for you to be able to fix it?

cc @cpuhrsch

**Comment by cpuhrsch** [comment]
@ezyang - that code is indeed responsible for that, but it's been updated and moved around since I first created it, so a more recent author should have better insight. Having said that, I've never encountered this behavior. Could this be something specific to Windows?

Also cc @colesbury @VitalyFedyunin

**Comment by peterjc123** [comment]
For in-source builds, `${CMAKE_BINARY_DIR}` is the equivalent of `${CMAKE_SOURCE_DIR}`. Are we doing in-source builds on systems other than Windows?

**Comment by peterjc123** [comment]
Tried out-source builds, failing with https://github.com/pytorch/pytorch/issues/20642.

**Comment by peterjc123** [comment]
Now it is failing with `lib\caffe2.lib : fatal error LNK1248: image size (1000EAAC7) exceeds maximum allowable size (FFFFFFFF)`. Looks like the unification of Caffe2 and PyTorch is breaking it.

**Comment by ezyang** [comment]
@peterjc123 Our Linux build is easily less than a gigabyte (based on the size of the downloads from pytorch.org nightlies webpage). Do you have any idea why your particular build is taking 2G?

**Comment by ezyang** [comment]
For reference, caffe2.dll from debug nightlies is 200M...

**Comment by peterjc123** [comment]
@ezyang Caffe2 Windows builds is started with `BUILD_SHARED_LIB=OFF`, that way, the static library `caffe2.lib(a)` is generated instead of `caffe2.dll(.so)`. Do we want to change that?

**Comment by Silas606** [comment]
I believe both libtorch and pytorch have `BUILD_SHARED_LIB=OFF` and out-source build as default so I think it is good idea to align caffe2 with them by default if there is no other specific reasons.

**Comment by peterjc123** [comment]
PyTorch uses `BUILD_SHARED_LIB=ON` and out-source build currently, and Caffe2 does the opposite. I'm okay with out-source build. However, for `BUILD_SHARED_LIB=OFF`, if we don't test them through CI, then they will possibly break in the future.

**Comment by soumith** [comment]
if for Windows you have the limit of `BUILD_SHARED_LIB=OFF` going above size limit, i think it's fine on Windows to default it to `ON`. And as you suggested, we should check this with CI, so add a CI entry to test that.

**Comment by peterjc123** [comment]
The new error comes.
```
  CMake Error at cmake/Modules_CUDA_fix/upstream/FindCUDA.cmake:1616 (configure_file):
    configure_file Problem configuring file
  Call Stack (most recent call first):
    cmake/Modules_CUDA_fix/upstream/FindCUDA.cmake:1830 (CUDA_WRAP_SRCS)
    cmake/public/utils.cmake:158 (cuda_add_library)
    caffe2/CMakeLists.txt:982 (torch_cuda_based_add_library)
```

**Comment by peterjc123** [comment]
Well, I found a related issue: https://gitlab.kitware.com/cmake/cmake/issues/19303. I will try to regenerate new AMI to update cmake and VS 2019.

**Comment by ezyang** [comment]
cc @magicwyzh @Xfel @HerculesJack

**Comment by ezyang** [comment]
@peterjc123 I'm going to de-prioritize this particular issue as it was caused by in-source build. I guess in principle our cmake could be made to also work with non-in-source build, but making this work feels like low priority to me.

**Comment by peterjc123** [comment]
@ezyang Yes, that seems okay to me. I'm just trying to moving all the builds to out-source builds to avoid this issue.

**Comment by anjali411** [comment]
Making this high priority because my build is failing when I try to add a new file with function at::native::computeStrideForComplex with this error which seems related to this issue.
```UnaryOpsKernel.cpp.AVX.cpp:(.text+0x483f0): multiple definition of `at::native::computeStrideForComplex(c10::ArrayRef<long>)'
caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp.AVX2.cpp.o:UnaryOpsKernel.cpp.AVX2.cpp:(.text+0x461a0): first defined here
caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp.DEFAULT.cpp.o: In function `at::native::computeStrideForComplex(c10::ArrayRef<long>)':
UnaryOpsKernel.cpp.DEFAULT.cpp:(.text+0x4ab00): multiple definition of `at::native::computeStrideForComplex(c10::ArrayRef<long>)'
caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp.AVX2.cpp.o:UnaryOpsKernel.cpp.AVX2.cpp:(.text+0x461a0): first defined here```

**Comment by ezyang** [comment]
@anjali411 Are you building in-source? If you're using `setup.py` that shouldn't be the case, so I don't see how you could be triggering this problem.

**Comment by VitalyFedyunin** [comment]
Unlabeling as more info needed

**Comment by anjali411** [comment]
> @anjali411 Are you building in-source? If you're using `setup.py` that shouldn't be the case, so I don't see how you could be triggering this problem.

yeah I am
